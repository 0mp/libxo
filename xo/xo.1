.\" #
.\" # Copyright (c) 2014, Juniper Networks, Inc.
.\" # All rights reserved.
.\" # This SOFTWARE is licensed under the LICENSE provided in the
.\" # ../Copyright file. By downloading, installing, copying, or 
.\" # using the SOFTWARE, you agree to be bound by the terms of that
.\" # LICENSE.
.\" # Phil Shafer, July 2014
.\" 
.Dd July, 2014
.Dt LIBXO 3
.Os
.Sh NAME
.Nm xo
.Nd emit formatted output based on format string and arguments
.Sh SYNOPSIS
.Nm xo
.Op Fl options
.Op Ar argument...
.Sh DESCRIPTION
The
.Nm xo
utility allows command line access to the functionality of
the
.Em libxo
library.  Using
.Nm xo ,
shell scripts can emit
.Em XML ,
.Em JSON , or
.Em HTML
using the same commands that emit text output.
.Pp
The style of output can be selected using a specific option:
.Fl X
for XML,
.Fl J
for JSON,
.Fl -H
for HTML, or
.Fl -T
for TEXT, which is the
default.  The
.Fl "-style <style>"
option can also be used.  The
.Em LIBXO_OPTIONS
environment variable can also be used to set the style,
as well as other flags.

The
.Nm xo
utility accepts a format string suitable for
.Xr xo_emit 3
and a set of zero or more arguments used to supply data for that string.
.Bd -literal -offset indent
    xo "The {k:name} weighs {:weight/%d} pounds.\n" fish 6

  TEXT:
    The fish weighs 6 pounds.
  XML:
    <name>fish</name>
    <weight>6</weight>
  JSON:
    "name": "fish",
    "weight": 6
  HTML:
    <div class="line">
      <div class="text">The </div>
      <div class="data" data-tag="name">fish</div>
      <div class="text"> weighs </div>
      <div class="data" data-tag="weight">6</div>
      <div class="text"> pounds.</div>
    </div>
.Ed
.Pp
The
.Fl "-wrap <path>"
option can be used to wrap emitted content in a
specific hierarchy.  The path is a set of hierarchical names separated
by the '/' character.
.Bd -literal -offset indent
    xo --wrap top/a/b/c '{:tag}' value

  XML:
    <top>
      <a>
        <b>
          <c>
            <tag>value</tag>
          </c>
        </b>
      </a>
    </top>
  JSON:
    "top": {
      "a": {
        "b": {
          "c": {
            "tag": "value"
          }
        }
      }
    }
.Ed
.Pp
The
.Fl "\-open <path>"
and
.Fl "\-close <path>"
can be used to emit
hierarchical information without the matching close and open
tag.  This allows a shell script to emit open tags, data, and
then close tags.  The
.Fl \-depth
option may be used to set the
depth for indentation.  The
.Fl "\-leading-xpath"
may be used to
prepend data to the XPath values used for HTML output style.
.Bd -literal -offset indent
    #!/bin/sh
    xo --open top/data
    xo --depth 2 '{tag}' value
    xo --close top/data
  XML:
    <top>
      <data>
        <tag>value</tag>
      </data>
    </top>
  JSON:
    "top": {
      "data": {
        "tag": "value"
      }
    }
.Ed
.Pp
.Sh COMMAND LINE OPTIONS
.Bd -literal -offset indent
Usage: xo [options] format [fields]
  --close <path>        Close tags for the given path
  --depth <num>         Set the depth for pretty printing
  --help                Display this help text
  --html OR -H          Generate HTML output
  --json OR -J          Generate JSON output
  --leading-xpath <path> Add a prefix to generated XPaths (HTML)
  --open <path>         Open tags for the given path
  --pretty OR -p        Make 'pretty' output (add indent, newlines)
  --style <style>       Generate given style (xml, json, text, html)
  --text OR -T          Generate text output (the default style)
  --version             Display version information
  --warn OR -W          Display warnings in text on stderr
  --warn-xml            Display warnings in xml on stdout
  --wrap <path>         Wrap output in a set of containers
  --xml OR -X           Generate XML output
  --xpath               Add XPath data to HTML output);
.Ed
.Pp
.Sh EXAMPLE
.Bd -literal -offset indent
  % xo 'The {:product} is {:status}\n' stereo "in route"
  The stereo is in route
  % xo -p -X 'The {:product} is {:status}\n' stereo "in route"
  <product>stereo</product>
  <status>in route</status>
.Ed
.Pp
.Sh ADDITIONAL DOCUMENTATION
.Pp
Complete documentation can be found on github:
.Bd -literal -offset indent
http://juniper.github.io/libxo/libxo-manual.html
.Ed
.Pp
libxo lives on github as:
.Bd -literal -offset indent
https://github.com/Juniper/libxo
.Ed
.Pp
The latest release of libxo is available at:
.Bd -literal -offset indent
https://github.com/Juniper/libxo/releases
.Ed
.Sh SEE ALSO
.Xr xo_emit 3
.Sh HISTORY
The
.Fa libxo
library was added in FreeBSD 10.1.
.Sh AUTHOR
Phil Shafer
