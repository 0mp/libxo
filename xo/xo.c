/*
 * Copyright (c) 2014, Juniper Networks, Inc.
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * ../Copyright file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, July 2014
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>

#include "libxo.h"
#include "xoversion.h"

#ifndef UNUSED
#define UNUSED __attribute__ ((__unused__))
#endif /* UNUSED */

#ifndef HAVE_STREQ
static inline int
streq (const char *red, const char *blue)
{
    return (red && blue && *red == *blue && strcmp(red + 1, blue + 1) == 0);
}
#endif /* HAVE_STREQ */

static int opt_warn;		/* Enable warnings */

static void
print_help (void)
{
    fprintf(stderr, "Usage: xo [options] formt [fields]\n");
}

static void
print_version (void)
{
    printf("libxo version %s%s\n", LIBXO_VERSION, LIBXO_VERSION_EXTRA);
}

static char *
check_arg (const char *name, char ***argvp, int has_opt)
{
    char *opt = NULL, *arg;

    if (has_opt) {
	opt = **argvp;
	*argvp += 1;
    }
    arg = **argvp;
    if (!has_opt)
	*argvp += 1;

    if (arg == NULL) {
	if (has_opt)
	    xo_error("missing %s argument for '%s' option", name, opt);
	else
	    xo_error("missing argument for '%s'", name);
	exit(1);
    }

    return arg;
}

static char **save_argv;

static char *
next_arg (void)
{
    char *cp = *save_argv;

    if (cp == NULL) {
	xo_error(NULL, "missing argument\n");
	exit(1);
    }

    save_argv += 1;
    return cp;
}

static void
prep_arg (char *fmt)
{
    char *cp, *fp;

    for (cp = fp = fmt; *cp; cp++, fp++) {
	if (*cp != '\\') {
	    if (cp != fp)
		*fp = *cp;
	    continue;
	}

	switch (*++cp) {
	case 'n':
	    *fp = '\n';
	    break;

	case 'r':
	    *fp = '\r';
	    break;

	case 'b':
	    *fp = '\b';
	    break;

	case 'e':
	    *fp = '\e';
	    break;

	default:
	    *fp = *cp;
	}
    }

    *fp = '\0';
}

/*
 * Our custom formatter is responsible for combining format string pieces
 * with our command line arguments to build strings.  This involves faking
 * some printf-style logic.
 */
static int
formatter (xo_handle_t *xop, xchar_t *buf, int bufsiz,
	   const xchar_t *fmt, va_list vap UNUSED)
{
    int lflag = 0, hflag = 0, jflag = 0, tflag = 0,
	zflag = 0, qflag = 0, star1 = 0, star2 = 0;
    int rc = 0;
    int w1 = 0, w2 = 0;
    const xchar_t *cp;

    for (cp = fmt + 1; *cp; cp++) {
	if (*cp == 'l')
	    lflag += 1;
	else if (*cp == 'h')
	    hflag += 1;
	else if (*cp == 'j')
	    jflag += 1;
	else if (*cp == 't')
	    tflag += 1;
	else if (*cp == 'z')
	    zflag += 1;
	else if (*cp == 'q')
	    qflag += 1;
	else if (*cp == '*') {
	    if (star1 == 0)
		star1 = 1;
	    else
		star2 = 1;
	} else if (strchr("diouxXDOUeEfFgGaAcCsSp", *cp) != NULL)
	    break;
	else if (*cp == 'n' || *cp == 'v') {
	    if (opt_warn)
		xo_error_h(xop, "unsupported format: '%s'", fmt);
	    return -1;
	}
    }

    xchar_t fc = *cp;

    /* Handle "%*.*s" */
    if (star1)
	w1 = strtol(next_arg(), NULL, 0);
    if (star2 > 1)
	w2 = strtol(next_arg(), NULL, 0);

    if (fc == 'D' || fc == 'O' || fc == 'U')
	lflag = 1;

    if (strchr("diD", fc) != NULL) {
	long long value = strtoll(next_arg(), NULL, 0);
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);

    } else if (strchr("ouxXOUp", fc) != NULL) {
	unsigned long long value = strtoull(next_arg(), NULL, 0);
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);

    } else if (strchr("eEfFgGaA", fc) != NULL) {
	double value = strtold(next_arg(), NULL);
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);

    } else if (fc == 'C' || fc == 'c' || fc == 'S' || fc == 's') {
	char *value = next_arg();
	if (star1 && star2)
	    rc = snprintf(buf, bufsiz, fmt, w1, w2, value);
	else if (star1)
	    rc = snprintf(buf, bufsiz, fmt, w1, value);
	else
	    rc = snprintf(buf, bufsiz, fmt, value);
    }

    return rc;
}

int
main (int argc UNUSED, char **argv)
{
    char *fmt = NULL, *cp, *np, *container = NULL;
    int clen = 0;

    for (argv++; *argv; argv++) {
	cp = *argv;

	if (*cp != '-')
	    break;

	if (streq(cp, "--"))
	    break;

	if (streq(cp, "--help")) {
	    print_help();
	    return 1;

	} else if (streq(cp, "--container") || streq(cp, "-c")) {
	    container = check_arg("container", &argv, 1);

	} else if (streq(cp, "--html") || streq(cp, "-H")) {
	    xo_set_style(NULL, XO_STYLE_HTML);

	} else if (streq(cp, "--json") || streq(cp, "-J")) {
	    xo_set_style(NULL, XO_STYLE_JSON);

        } else if (streq(cp, "--pretty") || streq(cp, "-p")) {
	    xo_set_flags(NULL, XOF_PRETTY);

	} else if (streq(cp, "--style") || streq(cp, "-s")) {
	    np = check_arg("style", &argv, 1);

	    if (streq(cp, "xml"))
		xo_set_style(NULL, XO_STYLE_XML);
	    else if (streq(cp, "json"))
		xo_set_style(NULL, XO_STYLE_JSON);
	    else if (streq(cp, "text"))
		xo_set_style(NULL, XO_STYLE_TEXT);
	    else if (streq(cp, "html"))
		xo_set_style(NULL, XO_STYLE_HTML);
	    else {
		xo_error("unknown style: %s", cp);
		exit(1);
	    }

	} else if (streq(cp, "--text") || streq(cp, "-T")) {
	    xo_set_style(NULL, XO_STYLE_TEXT);

	} else if (streq(cp, "--xml") || streq(cp, "-X")) {
	    xo_set_style(NULL, XO_STYLE_XML);

	} else if (streq(cp, "--xpath")) {
	    xo_set_flags(NULL, XOF_XPATH);

	} else if (streq(cp, "--version")) {
	    print_version();
	    return 0;

	} else if (streq(cp, "--warn") || streq(cp, "-W")) {
	    opt_warn = 1;
	    xo_set_flags(NULL, XOF_WARN);
	}
    }

    fmt = check_arg("format string", &argv, 0);
    if (fmt == NULL || *fmt == '\0')
	return 0;

    prep_arg(fmt);
    xo_set_formatter(NULL, formatter);
    xo_set_flags(NULL, XOF_NO_VA_ARG);

    save_argv = argv;

    if (container) {
	clen = strlen(container);
	for (cp = container; cp && *cp; cp = np) {
	    np = strchr(cp, '/');
	    if (np)
		*np = '\0';
	    xo_open_container(cp);
	    if (np)
		*np++ = '/';
	}
    }

    xo_emit(fmt);

    while (container) {
	np = strrchr(container, '/');
	xo_close_container(np ? np + 1 : container);
	if (np)
	    *np = '\0';
	else
	    container = NULL;
    }

    xo_flush();

    return 0;
}
