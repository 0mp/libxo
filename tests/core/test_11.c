/*
 * Copyright (c) 2014, Juniper Networks, Inc.
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * ../Copyright file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, July 2014
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <ctype.h>
#include <syslog.h>

#include "xo.h"

void
test_syslog_open (void)
{
    printf("syslog open\n");
}

void
test_syslog_close (void)
{
    printf("syslog close\n");
}

void
test_syslog_send (const char *full_msg, const char *v0_hdr,
		  const char *text_only)
{
    printf("{{%s}}\n{{%s}}\n{{%s}}\n\n", full_msg, v0_hdr, text_only);
}

int
main (int argc, char **argv)
{
    
    argc = xo_parse_args(argc, argv);
    if (argc < 0)
	return 1;

    xo_set_syslog_handler(test_syslog_open, test_syslog_send,
			  test_syslog_close);

    xo_set_unit_test_mode(1);
    xo_open_log("test-program", LOG_PERROR, 0);

    xo_set_version("3.1.4");
    xo_set_syslog_enterprise_id(42); /* SunOs */

    xo_open_container_h(NULL, "top");

    xo_syslog(LOG_INFO | LOG_KERN, "animal-status",
	      "The {:animal} is {:state}", "snake", "loose");
    xo_syslog(LOG_INFO | LOG_MAIL, "animal-consumed",
	      "My {:animal} ate your {:pet}", "snake", "hamster");
    xo_syslog(LOG_NOTICE | LOG_DAEMON, "animal-talk",
	      "{:count/%d} {:animal} said {:quote}", 1, "owl", "\"e=m\\c[2]\"");

    xo_close_container_h(NULL, "top");

    xo_finish();

    return 0;
}
