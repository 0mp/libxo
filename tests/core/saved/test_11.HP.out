{{<14>1 2015-06-23T18:47:09.000-0500 bock.juniper.net test-program 222 [42@animal-status <div class="line">
  <div class="text">The </div>
  <div class="data" data-tag="animal">snake</div>
  <div class="text"> is </div>
  <div class="data" data-tag="state">loose</div>
</div>
] The snake is loose}}
{{test-program: }}
{{The snake is loose}}

{{<22>1 2015-06-23T18:47:09.000-0500 bock.juniper.net test-program 222 [42@animal-consumed <div class="line">
  <div class="text">My </div>
  <div class="data" data-tag="animal">snake</div>
  <div class="text"> ate your </div>
  <div class="data" data-tag="pet">hamster</div>
</div>
] My snake ate your hamster}}
{{test-program: }}
{{My snake ate your hamster}}

{{<29>1 2015-06-23T18:47:09.000-0500 bock.juniper.net test-program 222 [42@animal-talk <div class="line">
  <div class="data" data-tag="count">1</div>
  <div class="text"> </div>
  <div class="data" data-tag="animal">owl</div>
  <div class="text"> said </div>
  <div class="data" data-tag="quote">"e=m\c[2]"</div>
</div>
] 1 owl said "e=m\c[2]"}}
{{test-program: }}
{{1 owl said "e=m\c[2]"}}

